File: solutions/328. Odd Even Linked List.cpp
using Node = ListNode;
class Solution {
public:
    ListNode* oddEvenList(ListNode* root) {
        if (root == nullptr || root->next == nullptr) {
            return root;
        }
        Node *even = nullptr, *odd = nullptr, *tmp = root;
        Node* t = nullptr, *p = nullptr; int c = 1;
        while(tmp != nullptr) {
            if (c & 1) {
                if (odd == nullptr) {
                    odd = new Node(tmp->val);
                    t = odd;
                } else {
                    Node *newnode = new Node(tmp->val);
                    odd->next = newnode;
                    odd = newnode;
                }
            } else {
                if (even == nullptr) {
                    even = new Node(tmp->val);
                    p = even;
                }else {
                    Node *newnode = new Node(tmp->val);
                    even->next = newnode;
                    even = newnode;
                }
            }
            tmp = tmp->next;
            c++;
        }
        odd->next = p;
        root = t;
        return root;
    }
};

File: solutions/328. Odd Even Linked List.java
class Solution {
    public ListNode oddEvenList(ListNode head) {
        ListNode doddIndices = new ListNode(-1);
        ListNode devenIndices = new ListNode(-1);
​
        ListNode oddIndices = doddIndices;
        ListNode evenIndices = devenIndices;
        ListNode node = head;
        int index = 0;
        while (node != null) {
            boolean isEven = (index % 2 == 0);
            if (isEven) {
                evenIndices.next = new ListNode(node.val);
                evenIndices = evenIndices.next;
            } else {
                oddIndices.next = new ListNode(node.val);
                oddIndices = oddIndices.next;
            }
            node = node.next;
            index++;
        }
        evenIndices.next = doddIndices.next;
        return devenIndices.next;
    }
}
